### PR TITLE
reboot fix Issue 23058 - importC: cannot take address inside multi-di…

### PR DESCRIPTION
**Pull request by WalterBright**
…mensional array at compile time

I was forced to abandon https://github.com/dlang/dmd/pull/14044 as it capsized and sank. I wish git had an undo. Sure would be nice.

### CODE REVIEW

File: test/fail_compilation/test23058.c [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+/* https://issues.dlang.org/show_bug.cgi?id=23058
+ */
+
+#line 100
+
+int arr[3][4] = { { 1,2,3,4 }, { 5,6,7,8 }, { 9,10,11,12 } };
+
+void test1()
+{
+    int *p1 = &arr[4][2];
```
Comment (RazvanN7) [inline test/fail_compilation/test23058.c:17]: This code is technically correct code from the perspective of the C specification. Is it ok to turn it into an error?

Comment (WalterBright) [reply test/fail_compilation/test23058.c:17]: It's length+1, which is not technically correct as I recall.

Comment (RazvanN7) [reply test/fail_compilation/test23058.c:17]: I compiled this code with gcc and there's no error. Also, the actual code [you are introducing](https://github.com/dlang/dmd/pull/14094/files#diff-be3f929dc15535ccf5bbb3d9f274993cac59cad6f83105dda698b28338bbb323R571) treats C and D code the same.

File: test/fail_compilation/fail13902.d [hunk 2]
```diff
@@ -57,7 +57,7 @@ fail_compilation/fail13902.d(88): Error: Using the result of a comma expression
 fail_compilation/fail13902.d(75): Error: returning `& x` escapes a reference to parameter `x`
 fail_compilation/fail13902.d(76): Error: returning `&s1.v` escapes a reference to parameter `s1`
 fail_compilation/fail13902.d(81): Error: returning `& sa1` escapes a reference to parameter `sa1`
-fail_compilation/fail13902.d(82): Error: returning `&sa2[0][0]` escapes a reference to parameter `sa2`
+fail_compilation/fail13902.d(82): Error: returning `& sa2` escapes a reference to parameter `sa2`
```
Comment (SixthDot) [inline test/fail_compilation/fail13902.d:60]: off topic: why does `&` is formatted sometimes with a trailing space and sometimes not ?

Comment (dkorpel) [reply test/fail_compilation/fail13902.d:60]: The parser generates an `AddrExp` for `&x`, which gets printed as `&x` as well. Semantic analysis turn an `AddrExp` it into a `SymOffExp` with offset 0, which gets printed as `& x`. (See dmd.hdrgen.d for the relevant source code)

Comment (SixthDot) [reply test/fail_compilation/fail13902.d:60]: shouldn't this be considered as an issue ? What's the point ?

Comment (SixthDot) [reply test/fail_compilation/fail13902.d:60]: I mean, why the space ?

Comment (dkorpel) [reply test/fail_compilation/fail13902.d:60]: I don't know why the space was added, but from a compiler developer perspective, it is useful to distinguish an `AddrExp` or `SymOff` expression.

Comment (SixthDot) [reply test/fail_compilation/fail13902.d:60]: that space should be removed  IMO.

File: test/fail_compilation/test23058.c (hunk 1 above)
Comment (WalterBright) [reply test/fail_compilation/test23058.c:17]: You're right that gcc doesn't give an error for:
```
int a[10];
int* f() { return &a[100]; }
```
but clang does with `-Wpedantic -std=c11`
The C11 6.5.6-8 spec lists such constructs as "undefined behavior". I'm good with diagnosing an error at compile time.

Comment (WalterBright) [reply test/fail_compilation/test23058.c:17]: There's no reason it should be different for D.

**Comment by WalterBright** [comment]
@RazvanN7 ?

**Review by dkorpel — APPROVED** [review]

**Comment by ibuclaw** [comment]
This PR introduced weird codegen for D array indexing.
```
auto test(const(ubyte[32])[] a)
{
    return cast(const(ubyte)*)&a[1][0];
}
```

Before this, frontend passed to the code generator:
```
return &a[1][0];
```
With this change, we now see:
```
return &a[1] + cast(const(ubyte)*)0LU;
```

File: src/dmd/optimize.d [hunk 3]
```diff
@@ -558,6 +558,41 @@ Expression Expression_optimize(Expression e, int result, bool keepLvalue)
...
+                    if (overflow)
+                    {
+                        e.error("array offset overflow");
+                        return error();
+                    }
+
+                    Expression ex = new AddrExp(ae1.loc, ae1);  // &a[i]
+                    ex.type = ae1.type.pointerTo();
+
+                    Expression add = new AddExp(ae.loc, ex, new IntegerExp(ae.loc, offset, e.type));
```
Comment (ibuclaw) [inline src/dmd/optimize.d:589]: I assume this is the problem - the integer expression is given a pointer type.

**Pull request merged**